src/utils/text-measurement.js

```diff
@@ -76,14 +76,34 @@ class TextMeasurement {
 
     // Approximate the number of characters to truncate to,
     // using avg character width as reference.
-    const f = (maxWidth - this.minWidth) / this._averageCharWidth;
+    const availableWidth = maxWidth - this.minWidth;
+    const f = availableWidth / this._averageCharWidth;
     let n = Math.floor(f);
-    if (n === f) {
-      // The approximate width of `n` characters is exactly max width,
-      // so take one character less just in case.
-      n -= 1;
+    if (n < 1) {
+      return '';
     }
-    return n > 0 ? text.substring(0, n) + this.overflowChar : '';
+
+    const getWidth = () => this.getTextWidth(text.substring(0, n));
+    const getRemainingCharacterCount = () =>
+      Math.round((availableWidth - getWidth()) / this._averageCharWidth);
+
+    // Do a second finer grained approximation to add or remove a few characters.
+    n += getRemainingCharacterCount();
+
+    // And a third one that can only add characters. This will be useful when
+    // the characters added at the previous step were narrow (eg. '::').
+    const offset = getRemainingCharacterCount();
+    if (offset >= 1) {
+      n += offset;
+    }
+
+    // If we overflow a little bit, remove characters one at a time until we no
+    // longer do.
+    while (availableWidth - getWidth() < 0) {
```

Comment:
I'd find this easier to read if it were `while (getWidth(n) > availableWidth) {`